core/trino-main/src/main/java/io/trino/operator/PartitionedTableWriterOperator.java

```diff
@@ -0,0 +1,580 @@
+        for (int i = 0; i < partitionAssignments.length; i++) {
+            partitionAssignments[i] = new IntArrayList();
+            partitionPageSinks[i] = null;
+        }
+
+        this.tableWriterInfoSupplier = createTableWriterInfoSupplier(pageSinkPeakMemoryUsage, statisticsTiming, partitionPageSinks);
+        this.operatorContext.setInfoSupplier(tableWriterInfoSupplier);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public void finish()
+    {
+        OperationTimer timer = new OperationTimer(statisticsCpuTimerEnabled);
+        statisticAggregationOperator.finish();
+        timer.end(statisticsTiming);
+
+        ImmutableList.Builder<ListenableFuture<?>> blockedFutures = ImmutableList.builder();
+        blockedFutures.add(blocked);
+        blockedFutures.add(statisticAggregationOperator.isBlocked());
+
+        if (state == State.RUNNING) {
+            state = State.FINISHING;
+            ImmutableList.Builder<CompletableFuture<Collection<Slice>>> futures = ImmutableList.builder();
+            for (ConnectorPageSink pageSink : partitionPageSinks) {
+                if (pageSink != null) {
+                    CompletableFuture<Collection<Slice>> finish = pageSink.finish();
+                    futures.add(finish);
+                    blockedFutures.add(toListenableFuture(finish));
+                }
+            }
+            pageSinkFinishFutures = futures.build();
+            updateWrittenBytes();
+        }
+        this.blocked = asVoid(allAsList(blockedFutures.build()));
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state == State.FINISHED && blocked.isDone();
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        return blocked;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        if (state != State.RUNNING || !blocked.isDone()) {
+            return false;
+        }
+        return statisticAggregationOperator.needsInput();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+        checkState(needsInput(), "Operator does not need input");
+
+        OperationTimer timer = new OperationTimer(statisticsCpuTimerEnabled);
+        statisticAggregationOperator.addInput(page);
+        timer.end(statisticsTiming);
+        ListenableFuture<Void> blockedOnAggregation = statisticAggregationOperator.isBlocked();
+
+        ImmutableList.Builder<ListenableFuture<?>> blockedOnWrites = ImmutableList.builder();
+        blockedOnWrites.add(blockedOnAggregation);
+
+        Page partitionedPage = partitionPagePreparer.apply(page);
+
+        // Assign each row to a partition which limits to the partitionCount. If there are more partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
```

Comment:
`If there are more partitions than` -> `If there are more physical partitions than.., then it's possible that multiple physical partitions`